File: 623-add-one-row-to-tree/623-add-one-row-to-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode tra(TreeNode r, int v, int d, int check){
        if(d==1){
            if(check==1)
                return new TreeNode(v,r,null);
            return new TreeNode(v,null,r);
        }
        if(r==null)
            return null;
        r.left=tra(r.left,v,d-1,1);
        r.right=tra(r.right,v,d-1,0);
        return r;
    }
    public TreeNode addOneRow(TreeNode root, int val, int depth) {
        return tra(root,val,depth,1);
    }
}

File: 623-add-one-row-to-tree/623-add-one-row-to-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* tra(TreeNode* r, int v, int d,int l_end, int r_end){
        if(d==1){
            if(l_end==1)
                return new TreeNode(v,r,NULL);
            if(r_end==1)
                return new TreeNode(v,NULL,r);
        }
        if(r==NULL){
           return NULL; 
        }
        r->left=tra(r->left,v,d-1,1,0);
        r->right=tra(r->right,v,d-1,0,1);
        return r;
        
    }
    TreeNode* addOneRow(TreeNode* root, int val, int depth) {
        return tra(root,val,depth,1,1);
    }
};